pkg/controller/garbagecollector/garbage_collector.go

```diff
@@ -0,0 +1,421 @@
+	event := &event{
+		eventType: deleteEvent,
+		obj:       objectReferenceToMetadataOnlyObject(identity),
+	}
+	glog.V(5).Infof("generating virtual delete event for %s\n\n", event.obj)
+	gc.dependencyGraphBuilder.enqueueChanges(event)
+}
+
+func ownerRefsToUIDs(refs []metav1.OwnerReference) []types.UID {
+	var ret []types.UID
+	for _, ref := range refs {
+		ret = append(ret, ref.UID)
+	}
+	return ret
+}
+
+func (gc *GarbageCollector) attempToDeleteItem(item *node) error {
+	glog.V(2).Infof("processing item %s", item.identity)
+	// "being deleted" is an one-way trip to the final deletion. We'll just wait for the final deletion, and then process the object's dependents.
+	if item.isBeingDeleted() && !item.isDeletingDependents() {
+		glog.V(5).Infof("processing item %s returned at once", item.identity)
+		return nil
+	}
+	// TODO: It's only necessary to talk to the API server if this is a
+	// "virtual" node. The local graph could lag behind the real status, but in
+	// practice, the difference is small.
+	latest, err := gc.getObject(item.identity)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// the GraphBuilder can add "virtual" node for an owner that doesn't
+			// exist yet, so we need to enqueue a virtual Delete event to remove
+			// the virtual node from GraphBuilder.uidToNode.
+			glog.V(5).Infof("item %v not found, generating a virtual delete event", item.identity)
+			gc.generateVirtualDeleteEvent(item.identity)
+			return nil
+		}
+		return err
+	}
+	if latest.GetUID() != item.identity.UID {
+		glog.V(5).Infof("UID doesn't match, item %v not found, generating a virtual delete event", item.identity)
+		gc.generateVirtualDeleteEvent(item.identity)
+		return nil
+	}
+
+	// TODO: attemptToOrphanWorker() routine is similar. Consider merging
+	// attemptToOrphanWorker() into attempToDeleteItem() as well.
+	if item.isDeletingDependents() {
+		return gc.processDeletingDependentsItem(item)
+	}
+
+	// compute if we should delete the item
+	ownerReferences := latest.GetOwnerReferences()
+	if len(ownerReferences) == 0 {
+		glog.V(2).Infof("object %s's doesn't have an owner, continue on next item", item.identity)
+		return nil
+	}
+
+	solid, dangling, waiting, err := gc.classifyReferences(item, ownerReferences)
+	if err != nil {
+		return err
+	}
+	glog.V(5).Infof("classify references of %s.\nsolid: %#v\ndangling: %#v\nwaiting: %#v\n", item.identity, solid, dangling, waiting)
+
+	switch {
+	case len(solid) != 0:
+		glog.V(2).Infof("object %s has at least one existing owner, will not garbage collect", item.identity)
+		if len(dangling) != 0 || len(waiting) != 0 {
+			glog.V(2).Infof("remove dangling references %#v and waiting references %#v for object %s", dangling, waiting, item.identity)
+		}
+		patch := deleteOwnerRefPatch(item.identity.UID, append(ownerRefsToUIDs(dangling), ownerRefsToUIDs(waiting)...)...)
+		_, err = gc.patchObject(item.identity, patch)
+		return err
+	case len(waiting) != 0 && len(item.dependents) != 0:
```

Comment:
It doesn't look like the  `len(item.dependents) != 0` is buying you anything.  Why is this different than just the default case since you'll just loop over nothing.